### PR TITLE
[Don't merge] debugging #32770

### PR DESCRIPTION
**Pull request by hongchaodeng**
ref: #32770

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/32948)

### CODE REVIEW

**Comment by hongchaodeng** [comment]
@k8s-bot unit test this issue: #IGNORE

**Comment by hongchaodeng** [comment]
I'm wondering if we could have some sort of debugging patches on top testing infra that's resettable.

**Comment by hongchaodeng** [comment]
@k8s-bot unit test this issue: #IGNORE

**Comment by hongchaodeng** [comment]
@k8s-bot unit test this issue: #IGNORE

**Comment by hongchaodeng** [comment]
@k8s-bot unit test this issue: #IGNORE

**Comment by hongchaodeng** [comment]
@k8s-bot unit test this issue: #IGNORE

**Comment by hongchaodeng** [comment]
@k8s-bot unit test this issue: #IGNORE

**Comment by hongchaodeng** [comment]
From the last failure logs:
```
label (foo=bar), action delete, takes 11.511711ms
received event: /registry/deployments/test/foo-deployment
received event: /registry/deployments/test/foo-deployment
received event: /registry/deployments/test/foo-deployment
received event: /registry/deployments/test/foo-deployment
exiting watching loop

field (metadata.name=foo-deployment), action create, takes 3.8113ms
exiting watching loop

field (metadata.name=foo-deployment), action delete, takes 12.160971ms
received event: /registry/deployments/test/foo-deployment
received event: /registry/deployments/test/foo-deployment
received event: /registry/deployments/test/foo-deployment
received event: /registry/deployments/test/foo-deployment
received event: /registry/deployments/test/foo-deployment
received event: /registry/deployments/test/foo-deployment
```
The action `field (metadata.name=foo-deployment), action create` isn't getting any event for 30s. But the action "create" is successful and event history is still correct from later logs.

@timothysc @wojtek-t @lavalamp @xiang90

**Comment by hongchaodeng** [comment]
@k8s-bot unit test this issue: #IGNORE

**Comment by hongchaodeng** [comment]
@k8s-bot unit test this issue: #IGNORE

**Comment by hongchaodeng** [comment]
@k8s-bot unit test this issue: #IGNORE

**Comment by hongchaodeng** [comment]
@k8s-bot unit test this issue: #IGNORE

**Comment by hongchaodeng** [comment]
@xiang90 @timothysc @wojtek-t @lavalamp 
I think I have very close to the root problem of this.
From the logs:
```
2016-09-18 15:31:43.644070 I | start watching: /registry/test/foo
2016-09-18 15:31:43.645432 I | exiting watching loop: /registry/test/foo
...
2016-09-18 15:31:43.661205 I | field (metadata.name=foo), action delete, takes 20.202864ms
2016-09-18 15:32:13.661863 I | timeout
```
This is the last watching of "/registry/test/foo" and last exiting of "/registry/test/foo".
The startWatching() loop sort of ends around "15:31:43". But it doesn't have any error and as a result the loop just stops silently. The test just times out while the other goroutines of the watcher keeps running.

#33003 should fix the problem. The code should be fixed anyway.
Meanwhile, I need to dig out what could lead to the such situation.

File: pkg/api/rest/resttest/resttest.go [hunk 1]
```diff
@@ -1286,16 +1292,19 @@ func (t *Tester) testWatchLabels(obj runtime.Object, emitFn EmitFunc, labelsPass
 			if err != nil {
 				t.Errorf("unexpected error: %v", err)
 			}
+			start := time.Now()
 			if err := emitFn(obj, action); err != nil {
 				t.Errorf("unexpected error: %v", err)
 			}
+			log.Printf("label (%v), action %v, takes %v", label, action, time.Since(start))
```
Comment (timothysc) [inline pkg/api/rest/resttest/resttest.go]: t.Logf() on this file.

**Comment by hongchaodeng** [comment]
@k8s-bot unit test this issue: #IGNORE

**Comment by hongchaodeng** [comment]
I can reproduce the problem pretty easily now with this patch and the script:
```bash
while [ 1 ] ; do
  go test -v ./pkg/registry/replicaset/etcd -run "TestWatch" -count 20 || break
  go test -v ./pkg/registry/configmap/etcd -run "TestWatch" -count 20 || break
  go test -v ./pkg/registry/deployment/etcd -run "TestWatch" -count 20 || break
  go test -v ./pkg/registry/namespace/etcd -run "TestWatch" -count 20 || break
  go test -v ./pkg/registry/networkpolicy/etcd -run "TestWatch" -count 20 || break
done
```
Run:
```
bash test.sh
```

**Comment by hongchaodeng** [comment]
I am able to reproduce it in pure etcd test.
Closing this.

**Comment by hongchaodeng** [comment]
kubemark test isn't relevant:
```
(gcloud.compute.firewall-rules.delete) Some requests did not succeed:
```

**Comment by hongchaodeng** [comment]
@k8s-bot test this issue: #IGNORE

**Pull request closed without merging**